swarm/src/connection.rs

```diff
@@ -244,6 +260,26 @@ where
                 Poll::Ready(ConnectionHandlerEvent::Close(err)) => {
                     return Poll::Ready(Err(ConnectionError::Handler(err)));
                 }
+                Poll::Ready(ConnectionHandlerEvent::ReportRemoteProtocols(
+                    ProtocolSupport::Added(protocols),
+                )) => {
+                    handler.on_connection_event(ConnectionEvent::RemoteProtocolsChange(
+                        ProtocolsChange::Added(ProtocolsAdded {
+                            protocols: protocols.difference(&HashSet::new()).peekable(), // This is a bit of a hack to use the same type internally in `ProtocolsAdded`.
+                        }),
+                    ));
+                    continue;
+                }
```

Comment:
I guess we could cache the remote protocols and then only pass a `ProtocolSupport::Added` back down to the `ConnectionHandler` in case it hasn't been reported before. 

Just an idea. Happy to go with at it is here.

Reply:
I've pushed a first version, will likely iterate on that further.